src/tss2-tcti/tcti-swtpm.c

```diff
@@ -56,18 +56,23 @@
 
 /*
  * This function wraps the "up-cast" of the opaque TCTI context type to the
- * type for the swtpm TCTI context. The only safeguard we have to ensure this
- * operation is possible is the magic number in the swtpm TCTI context.
- * If passed a NULL context, or the magic number check fails, this function
- * will return NULL.
+ * type for the mssim TCTI context. The only safeguard we have to ensure this
+ * operation is possible is the magic number in the mssim TCTI context.
+ * If passed a NULL context the function returns a NULL ptr
+ * If the magic number check fails, this function overwrites the magic with
+ * a bad number, which then should be caught by the tcti_common_check
+ * functions.
  */
 TSS2_TCTI_SWTPM_CONTEXT*
 tcti_swtpm_context_cast (TSS2_TCTI_CONTEXT *tcti_ctx)
 {
-    if (tcti_ctx != NULL && TSS2_TCTI_MAGIC (tcti_ctx) == TCTI_SWTPM_MAGIC) {
-        return (TSS2_TCTI_SWTPM_CONTEXT*)tcti_ctx;
-    }
-    return NULL;
+    if (tcti_ctx == NULL)
+	return NULL;
+
+    if (TSS2_TCTI_MAGIC (tcti_ctx) != TCTI_SWTPM_MAGIC)
+        TSS2_TCTI_MAGIC (tcti_ctx) = TCTI_BAD_MAGIC;
```

Comment:
Unfortunately this doesn't work for tpmclient because it [first tries to call `Tss2_Tcti_Swtpm_Reset`](https://github.com/tpm2-software/tpm2-tss/blob/3879a0b41bbecd9b4e229b3a73b6fd3835d0370a/test/tpmclient/tpmclient.int.c#L152) on the given TCTI context and [falls back to `tcti_platform_command(...,MS_SIM_POWER_OFF/ON)`](https://github.com/tpm2-software/tpm2-tss/blob/3879a0b41bbecd9b4e229b3a73b6fd3835d0370a/test/tpmclient/tpmclient.int.c#L161) in case that fails. This cast irrevocably destroys non-swtpm TCTI contexts if they are passed into swtpm TCTI functions, so they are unusable afterwards even when used with their appropriate TCTI functions. This seems dangerous because other applications might also expect these functions not to alter the passed TCTI context.

Reply:
Yes, but that's yet another problem, because tpm_client assumes that if TCTI_SWTPM is enabled then it can directly call Tss2_Tcti_Swtpm_Reset() on the tcti it gets, which doesn't necessarily need to be swtpm_tcti at all if we set TPM20_TCTI_NAME to mssim for instance.
The problem with the reset call is that it is not part of the TCTI interface as defined by the spec. The spec only defines these: https://github.com/tpm2-software/tpm2-tss/blob/master/include/tss2/tss2_tcti.h#L88-L132
In this case the ctx becomes invalid.

Reply:
Sure, you could avoid the problem by checking the magic number of the TCTI context before calling `Tss2_Tcti_Swtpm_Reset` in tpmclient. Still, I don't really see the purpose of invalidating the context: `TCTI_BAD_MAGIC` is never used for comparison, so what's the point?